prov/verbs/src/verbs_msg.c

```diff
@@ -39,40 +39,53 @@ static ssize_t
 fi_ibv_msg_ep_recvmsg(struct fid_ep *ep, const struct fi_msg *msg, uint64_t flags)
 {
 	struct fi_ibv_msg_ep *_ep;
-	struct ibv_recv_wr wr, *bad;
+	struct ibv_recv_wr *bad;
+	struct fi_ibv_wre *wre;
 	struct ibv_sge *sge = NULL;
 	ssize_t ret;
 	size_t i;
 
 	_ep = container_of(ep, struct fi_ibv_msg_ep, ep_fid);
 	assert(_ep->rcq);
 
-	wr.wr_id = (uintptr_t) msg->context;
-	wr.next = NULL;
+	wre = util_buf_alloc(_ep->rcq->wre_pool);
+	if (!wre)
+		return -FI_EAGAIN;
+	memset(wre, 0, sizeof(*wre));
+	wre->ep = _ep;
+
+	((struct fi_context *)msg->context)->internal[0] = wre;
+	wre->wr.rwr.wr_id = (uintptr_t)msg->context;
+	wre->wr.rwr.next = NULL;
```

Comment:
Can we not require FI_CONTEXT from the app? We could pass `(uintptr_t)wre` as `wr_id` in the request posted to verbs and store app's `context` with `wre` which can later be returned as part of completion.

Reply:
Thanks! It's fine point.